src/agent/proxy.c

```diff
@@ -0,0 +1,135 @@
+
+static const sd_bus_vtable proxy_job_vtable[] = {
+        SD_BUS_VTABLE_START(0), SD_BUS_METHOD("Ready", "s", "", proxy_job_method_set_ready, 0), SD_BUS_VTABLE_END
+};
+
+static int proxy_job_method_set_ready(sd_bus_message *m, void *userdata, UNUSED sd_bus_error *ret_error) {
+        const char *status = NULL;
+        int r = sd_bus_message_read(m, "s", &status);
+        if (r < 0) {
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_INVALID_ARGS, "Invalid arguments");
+        }
+
+        r = 0;
+        UnitActiveState active_state = active_state_from_string(status);
+        if (active_state == UNIT_ACTIVE) {
+                _cleanup_proxy_job_ ProxyJob *proxy = (ProxyJob *) userdata;
+                r = sd_bus_reply_method_return(proxy->request_message, "");
+        } else if (active_state == UNIT_FAILED) {
+                _cleanup_proxy_job_ ProxyJob *proxy = (ProxyJob *) userdata;
+                r = sd_bus_reply_method_errorf(
+                                proxy->request_message, SD_BUS_ERROR_FAILED, "Proxy service failed to start");
+        }
+
+        if (r < 0) {
+                hirte_log_errorf("Failed to sent ready status message to proxy: %s", strerror(-r));
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Failed to set ready status");
+        }
+
+        return sd_bus_reply_method_return(m, "");
+}
+
+
+ProxyJob *
+                proxy_job_new(Agent *agent,
+                              const char *node,
+                              const char *unit,
+                              const char *mode,
+                              sd_bus_message *request_message) {
+        static uint32_t next_id = 0;
+
+        _cleanup_proxy_job_ ProxyJob *proxy = malloc0(sizeof(ProxyJob));
+        if (proxy == NULL) {
+                return NULL;
+        }
+
+        proxy->ref_count = 1;
+        proxy->id = ++next_id;
+        proxy->agent = agent_ref(agent);
+        proxy->request_message = sd_bus_message_ref(request_message);
+        LIST_INIT(proxy_jobs, proxy);
+
+        proxy->node_name = strdup(node);
+        if (proxy->node_name == NULL) {
+                return NULL;
+        }
+        proxy->unit_name = strdup(unit);
+        if (proxy->unit_name == NULL) {
+                return NULL;
+        }
+        proxy->mode = strdup(mode);
+        if (proxy->mode == NULL) {
+                return NULL;
+        }
+
+        int r = asprintf(&proxy->object_path, "%s/%u", INTERNAL_PROXY_OBJECT_PATH_PREFIX, proxy->id);
+        if (r < 0) {
+                return NULL;
+        }
+
+        return steal_pointer(&proxy);
+}
+
+ProxyJob *proxy_job_ref(ProxyJob *proxy) {
+        proxy->ref_count++;
+        return proxy;
+}
+
+void proxy_job_unref(ProxyJob *proxy) {
+        proxy->ref_count--;
+        if (proxy->ref_count != 0) {
+                return;
+        }
+
+        LIST_REMOVE(proxy_jobs, proxy->agent->outstanding_proxy_jobs, proxy);
```

Comment:
This is not quite right LIST_REMOVE assumes the node is already in the list. This is not always the case at the moment. agent_method_create_proxy() adds it to the list, but not before doing some operations that can fail. If those fail the unref will try to remove the proxy even though its not in the list.

A better approach is to always handle the list add/remove in the same place, i.e. the agent. This means, that we remove the proxy from the list (and then unref it) in agent_method_remove_proxy()  and the final unref.

Reply:
This is related to the above weak ref comment.

Reply:
Additionally, this means we probably need to have an proxy_service_unexport() too, which we call when removing the proxy from the agent list.